### PR TITLE
latency based routing with circuit breaking on errors

### PR DESCRIPTION
**Pull request by bculberson**
Add new routing options to distribute query load on solr clusters.

### CODE REVIEW

**Comment by SendGridDX** [comment]
[CLA assistant check](https://cla.sendgrid.com/sendgrid/go-solr?pullRequest=49) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/go-solr?pullRequest=49) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/go-solr?pullRequest=49) it.</sub>

File: adaptiveRouter.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+func newLatencyHistory(uri string, recency int) *searchHistory {
+	return &searchHistory{uri: uri, timings: make([]time.Duration, recency), errors: make([]bool, recency)}
+}
+
+func (u *searchHistory) addSearchResult(timing time.Duration, error bool) {
+	u.timings[u.offset] = timing
+	u.errors[u.offset] = error
+	u.offset++
+	if u.offset == len(u.timings) {
```
Comment (Kane-Sendgrid) [inline adaptiveRouter.go]: Better to instantiate 0 length with N capacity, and then use cap(u.timings) for the check.

File: solr_http.go [hunk 2]
```diff
@@ -62,6 +63,9 @@ func (s *solrHttp) Update(nodeUris []string, singleDoc bool, doc interface{}, op
 	if len(nodeUris) == 0 {
 		return fmt.Errorf("[SolrHTTP] nodeuris: empty node uris is not valid")
 	}
+	if s.router != nil {
+		s.router.SortUris(nodeUris)
```
Comment (Kane-Sendgrid) [inline solr_http.go]: why return value is ignored?

**Comment by Kane-Sendgrid** [comment]
lgtm

**Comment by Kane-Sendgrid** [comment]
lgtm

**Comment by shawnfeldman** [comment]
lgtm once the build is passing

**Comment by Kane-Sendgrid** [comment]
lgtm

**Pull request merged**